README.md
Moolticute
==========

[![License](https://img.shields.io/badge/license-GPLv3%2B-blue.svg)](http://www.gnu.org/licenses/gpl.html)

[![Build Status](https://travis-ci.org/raoulh/moolticute.svg?branch=master)](https://travis-ci.org/raoulh/moolticute)

This project aims to be an easy companion to your [Mooltipass](http://www.themooltipass.com) device and extend
the power of the device to more platform/tools.
With it you can manage your Mooltipass with a crossplatform app, as well as provide a daemon service that
handle all USB communication with the device.

This tool is written with a daemon that runs in background, and a user interface app to control you Mooltipass.
Other clients could also connect and talk to the daemon (it uses a websocket connection and simple JSON messages).
The official Mooltipass App only works with Chrome as it relies on USB HID library that is only implemented in Chrome.
A Firefox (or any other browser) extension could easily be written by using the Moolticute daemon.

It is completely cross platform, and runs on Linux (using libusb), OS X (native IOKit API), and Windows (native HID API).

> Warning! This project is a work in progress!

### Downloads
Packages are build and available here: https://calaos.fr/mooltipass/

### Dependencies

##### Windows, Linux, OSX
 - Requires Qt 5.6 or higher.
 - Those Qt5 modules are required:
  - qt-base, qt-widgets, qt-gui, qt-network, qt-websockets

##### Linux
 - Requires libusb
 - Requires a [udev rule for libusb](https://github.com/bobsaintcool/mooltipass-udev)

##### Ubuntu 16.04
```bash
sudo apt install libqt5websockets5-dev libusb-dev libusb-1.0-0-dev qt-sdk qt5-qmake qt5-default
echo 'SUBSYSTEM=="usb", ATTRS{idVendor}=="16d0", ATTRS{idProduct}=="09a0", TAG+="uaccess"' | sudo tee /etc/udev/rules.d/50-mooltipass.rules
sudo udevadm control --reload-rules
```

##### Arch Linux
```bash
sudo pacman -S --needed qt5-websockets libusb qt5-base
echo 'SUBSYSTEM=="usb", ATTRS{idVendor}=="16d0", ATTRS{idProduct}=="09a0", TAG+="uaccess"' | sudo tee /etc/udev/rules.d/50-mooltipass.rules
sudo udevadm control --reload-rules
```

### How to build

For now, no binary releases are out yet. You will need to build the software by following the next step.

Two method can be used to build, by using QtCreator IDE, or from command line (typically on Linux). After the build succeeded, two executable are created:
 - moolticuted (the daemon process)
 - MoolticuteApp (the main GUI app)

##### Using Qt SDK

 - Download and install the Qt SDK from their [website](http://qt.io)
 - Start Qt-Creator
 - Open the main Moolticute.pro project file
 - Click on the "play" button to build and run

##### Command line

Qt needs to be installed correctly (see you Linux distribution for that)

```
mkdir build
cd build
qmake ../Moolticute.pro
make
```

Be careful to use the Qt5 qmake. You can check if you are using the correct qmake by using the command
```
➜  ~  qmake --version
QMake version 3.0
Using Qt version 5.5.1 in /usr/lib
```

On Gentoo, a wrapper is created for qmake, so this command should be used:
```
qmake -qt=5 ../Moolticute.pro
```

### Licensing

Moolticute is free software; you can redistribute it and/or modify it under the terms of the GNU Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version.




debian/changelog
moolticute (0.7.2) stable; urgency=medium

  * UNRELEASED

 -- root <root@f301d753b5d9>  Sat, 26 Aug 2017 06:23:52 +0000

moolticute (0.7.1) stable; urgency=medium

  * UNRELEASED

 -- root <root@29b4f4db746d>  Sat, 26 Aug 2017 05:46:29 +0000

moolticute (0.5.1-beta) stable; urgency=medium

  [ Raoul Hecky ]
  * Do not try to send something when no device is connected. Fix #14
  * Send error message when device is not connected
  * do not fail when description is not available
  * update for new toolchain and qt5.7.1
  * Fix qmake path for new toolchain

  [ raoulh ]
  * Qt5.6 requirement

  [ Raoul Hecky ]
  * Add 500mx delay at launch before starting MPManager
  * Add show_app API command

  [ Mathieu ]
  * exitting MMM when a command sent during MMM entering fails
  * more check for get cpz ctr
  * loading change numbers, serial number if fw > 1.2
  * loading serial number for mini v1.2 only
  * better debug for load parent & child nodes
  * adding node address to constructor,
  * adding duplicate lists, used for save packet generation
  * linked chain checking function
  * checking returns for get params, set date, get serial, get change number

  [ Raoul Hecky ]
  * rewrite findNodeWithAddressInList(), add const qualifier to MPNode
  * Fix return value if null
  * add default values to properties

  [ Quentin Bourgeois ]
  * <systemd/service: Add and install a basic systemd user unit file>
  * <Linux/XCB: Remove XCB dependencies for Linux builds>
  * <HttpClient: Remove unused include>
  * <udev: Add a udev rules that should benefit for systemd based system>
  * <README/ubuntu: On Ubuntu 16.04 its recommanded to use apt>

  [ Raoul Hecky ]
  * Fix build
  * Fix windows QApplication
  * Fix ifdef OSX/WIN, include of qt should be done first. fix #24
  * Fix bug when sending data node. EOD could be wrong in some cases

  [ Nicolas Aguirre ]
  * Code simplification for Paramters
  * Add define for all parameters defined in mooltipass fw
  * Add 6 more parameters to UI and server. New paramters are those defined with those following values

  [ Raoul Hecky ]
  * Finish work on missing paramaters from #13
  * Fix name of signals for new parameters
  * Display not implemented message with button import/export #13
  * Fix crash when cancel_request id is invalid and not request are running. #13
  * Display fw version in about #13
  * Fix param debug. Send only changed parameters to daemon
  * small refactor
  * Disconnect current job when it's done to prevent wrong signal being emited after
  * Implement UI for memcheck. core code can be easily integrated now.

  [ Quentin Bourgeois ]
  * Revert "<udev: Add a udev rules that should benefit for systemd based system>"
  * <systemd: Move the systemd service file systemd wide>

  [ Mathieu ]
  * successfully browsing through the memory
  * data reception speed diag
  * correct clearing of nodes lists, linked list loop detect
  * linked list check for data nodes
  * trigger full scan after receiving data start node
  * progress bar for integrity check, dedicated job for memory check and repair
  * checking favorites
  * function to generate save packets
  * generatesave: diffing the remaining items
  * following raoulh's comments
  * functions to repair linked list
  * removing iterator for favoriteaddrs
  * double checking DB when performing DB changes
  * using QDateTime::currentMSecsSinceEpoch() for diag purposes to stay compatible with qt5.6

  [ Raoul Hecky ]
  * Use QMetaEnum to get the text from enum

  [ Mathieu ]
  * setting jobs error messages for better user debug
  * first function to add orphan parent node
  * dedicated function for testing our db correction algos
  * starting to introduce virtual addresses, add orphan parent implemented
  * updating tagcheck to follow parent virtual addresses
  * correct set service function
  * function to replace virtual address with real addresses
  * addorphan for data nodes, new checks for data node integrity check

  [ Raoul Hecky ]
  * Fix some small bugs from #13
  * Display serial number in about. #13

  [ Mathieu ]
  * following raoul's comments

  [ Corentin Jabot ]
  * Fix compilation with VS2015
  * On windows, require a double-click to show the window from the systray
  * Mingw expects lowercase librairie names
  * Improve the device detection code on linux.
  * Fix a crash when getting a device description
  * Implement WSClient::isMPMini()
  * Show the device's state in the systray's tooltip.
  * Allow changing settings while no card is inserted.
  * Do not display the setting tab when no device is connected
  * Add a tab that is displayed when the deamon is not connected
  * Only hide the application window when launched automatically.
  * Replace the top-right icon
  * Load the device settings as soon as it is connected
  * Fix inverted no deamon/no device tabs
  * Do not display the serial when it is null
  * Add ui elements for uid & request & advanced settings
  * Wire the 'random starting pin' setting
  * Wire the 'Display Hash' setting
  * Wire the 'Lock/Unlock mode' setting
  * Improve the way the combo box are set
  * Implement the UID request feature.
  * Fix compilation on windows
  * Fix warning
  * Fix The registery value for auto startup on windows
  * Improve usb device management on linux
  * Intoduce a QLineEdit-based widget for password input
  * Add a random password generation widget.
  * Start improving the credential managemzent UI
  * Improve connections detection
  * Quit credential management mode when deamon crash or disconnected
  * Complete rework of the credential management mode
  * Do not Capitalize the service name
  * Display a message box when swutching from an unedited credential
  * Fix creating empty row in the credential view
  * Do not needlessly promt for unsaved items user when leaving MMM
  * Preserve actions order when toogling password visibility
  * Fix regression when adding a credential

  [ Mathieu ]
  * bug fix for current linux ci

src/Common.h
/******************************************************************************
 **  Copyright (c) Raoul Hecky. All Rights Reserved.
 **
 **  Moolticute is free software; you can redistribute it and/or modify
 **  it under the terms of the GNU General Public License as published by
 **  the Free Software Foundation; either version 3 of the License, or
 **  (at your option) any later version.
 **
 **  Moolticute is distributed in the hope that it will be useful,
 **  but WITHOUT ANY WARRANTY; without even the implied warranty of
 **  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 **  GNU General Public License for more details.
 **
 **  You should have received a copy of the GNU General Public License
 **  along with Foobar; if not, write to the Free Software
 **  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 **
 ******************************************************************************/
#ifndef COMMON_H
#define COMMON_H

#include <QtCore>
#include <functional>
#include <memory>
#include <type_traits>

#define MOOLTIPASS_VENDORID     0x16D0
#define MOOLTIPASS_PRODUCTID    0x09A0
#define MOOLTIPASS_USAGE_PAGE   0xFF31
#define MOOLTIPASS_USAGE        0x74

#define MOOLTIPASS_FAV_MAX      14
#define MOOLTIPASS_ADDRESS_SIZE 4
#define MP_NODE_SIZE            132

#define MOOLTIPASS_BLOCK_SIZE   32
#define MOOLTIPASS_DESC_SIZE    23 //max size allowed for description

#define CMD_MAX_RETRY            3 //retry sending command before dropping
#define CMD_DEFAULT_TIMEOUT     0xFFAAFFAA
#define CMD_DEFAULT_TIMEOUT_VAL 3000 //3seconds default timeout

//Data node header size. It contains the size of data in 4 bytes Big endian
#define MP_DATA_HEADER_SIZE      4

#define MOOLTICUTE_DAEMON_PORT  30035

//Max file size for standard data file
#define MP_MAX_FILE_SIZE        1024 * 10
#define MP_MAX_SSH_SIZE         1024 * 50

#define MC_SSH_SERVICE          "Moolticute SSH Keys"

//shared memory size
#define SHMEM_SIZE              128 * 1024

#define qToChar(s) s.toLocal8Bit().constData()
#define qToUtf8(s) s.toUtf8().constData()

#define MOOLTICUTE_DAEMON_LOG_SOCK    "moolticuted_local_log_sock"

#define BITMAP_ID_OFFSET        128
#define ID_KEYB_EN_US_LUT       BITMAP_ID_OFFSET+18
#define ID_KEYB_FR_FR_LUT       BITMAP_ID_OFFSET+19
#define ID_KEYB_ES_ES_LUT       BITMAP_ID_OFFSET+20
#define ID_KEYB_DE_DE_LUT       BITMAP_ID_OFFSET+21
#define ID_KEYB_ES_AR_LUT       BITMAP_ID_OFFSET+22
#define ID_KEYB_EN_AU_LUT       BITMAP_ID_OFFSET+23
#define ID_KEYB_FR_BE_LUT       BITMAP_ID_OFFSET+24
#define ID_KEYB_PO_BR_LUT       BITMAP_ID_OFFSET+25
#define ID_KEYB_EN_CA_LUT       BITMAP_ID_OFFSET+26
#define ID_KEYB_CZ_CZ_LUT       BITMAP_ID_OFFSET+27
#define ID_KEYB_DA_DK_LUT       BITMAP_ID_OFFSET+28
#define ID_KEYB_FI_FI_LUT       BITMAP_ID_OFFSET+29
#define ID_KEYB_HU_HU_LUT       BITMAP_ID_OFFSET+30
#define ID_KEYB_IS_IS_LUT       BITMAP_ID_OFFSET+31
#define ID_KEYB_IT_IT_LUT       BITMAP_ID_OFFSET+32  // So it seems Italian keyboards don't have ~`
#define ID_KEYB_NL_NL_LUT       BITMAP_ID_OFFSET+33
#define ID_KEYB_NO_NO_LUT       BITMAP_ID_OFFSET+34
#define ID_KEYB_PO_PO_LUT       BITMAP_ID_OFFSET+35  // Polish programmer keyboard
#define ID_KEYB_RO_RO_LUT       BITMAP_ID_OFFSET+36
#define ID_KEYB_SL_SL_LUT       BITMAP_ID_OFFSET+37
#define ID_KEYB_FRDE_CH_LUT     BITMAP_ID_OFFSET+38
#define ID_KEYB_EN_UK_LUT       BITMAP_ID_OFFSET+39
#define ID_KEYB_CZ_QWERTY_LUT   BITMAP_ID_OFFSET+51
#define ID_KEYB_EN_DV_LUT       BITMAP_ID_OFFSET+52
#define ID_KEYB_FR_MAC_LUT      BITMAP_ID_OFFSET+53
#define ID_KEYB_FR_CH_MAC_LUT   BITMAP_ID_OFFSET+54
#define ID_KEYB_DE_CH_MAC_LUT   BITMAP_ID_OFFSET+55
#define ID_KEYB_DE_MAC_LUT      BITMAP_ID_OFFSET+56
#define ID_KEYB_US_MAC_LUT      BITMAP_ID_OFFSET+57

class QLocalServer;

#if QT_VERSION < 0x050700
//This declares a qAsConst implementation if we build with Qt < 5.7

template< class T >
using mc_remove_reference_t = typename std::remove_reference<T>::type;
template<class T>
mc_remove_reference_t<T> const& qAsConst(T&&t){return t;}
template<class T>
T const qAsConst(T&&t){return std::forward<T>(t);}
template<class T>
T const& qAsConst(T&t){return t;}
#endif

class Common
{
public:
    static void installMessageOutputHandler(QLocalServer *logServer = nullptr);

    Q_ENUMS(MPStatus)
    typedef enum
    {
        UnknownStatus = -1,
        NoCardInserted = 0,
        Locked = 1,
        Error2 = 2,
        LockedScreen = 3,
        Error4 = 4,
        Unlocked = 5,
        Error6 = 6,
        Error7 = 7,
        Error8 = 8,
        UnkownSmartcad = 9,
    } MPStatus;
    static QHash<MPStatus, QString> MPStatusUserString, MPStatusString;

    static Common::MPStatus statusFromString(const QString &st);

    static QDate bytesToDate(const QByteArray &d);
    static QByteArray dateToBytes(const QDate &dt);

    static QJsonArray bytesToJson(const QByteArray &data);
    static QJsonObject bytesToJsonObjectArray(const QByteArray &data);

    static bool isProcessRunning(qint64 pid);

    static QJsonObject readSharedMemory(QSharedMemory &sh);
    static bool writeSharedMemory(QSharedMemory &sh, const QJsonObject &o);

    //create a unique identifier and save it into a list
    static QString createUid(QString prefix = QString());
    //release the unique id and remove it from the list
    static void releaseUid(QString uid);

    //mask log by removing passwords and data from log
    static QString maskLog(const QString &rawJson);

    typedef enum
    {
        MP_Classic = 0,
        MP_Mini = 1,
    } MPHwVersion;


    enum class LockUnlockModeFeatureFlags : uint {
        Disabled          = 0,
        Password          = 0x01,
        SendEnter         = 0x02,
        Login             = 0x04,
        SendWin_L         = 0x08,
        SendCtrl_Alt_Del  = 0x10
    };

    //Credentials favorites
    enum FavoriteFlags {
        FAV_NOT_SET = -1,
        FAV_1,
        FAV_2,
        FAV_3,
        FAV_4,
        FAV_5,
        FAV_6,
        FAV_7,
        FAV_8,
        FAV_9,
        FAV_10,
        FAV_11,
        FAV_12,
        FAV_13,
        FAV_14,
    };
};

Q_DECLARE_METATYPE(Common::MPStatus)
Q_DECLARE_METATYPE(Common::MPHwVersion)

#define CSS_BLUE_BUTTON "QPushButton {" \
                            "color: #fff;" \
                            "background-color: #60B1C7;" \
                            "padding: 12px;" \
                            "border: none;" \
                        "}" \
                        "QPushButton:hover {" \
                            "background-color: #3d96af;" \
                        "}" \
                        "QPushButton:pressed {" \
                            "background-color: #237C95;" \
                        "}" \
                        "QPushButton:!enabled {" \
                            "background-color: #cee8ef;" \
                        "}"

#define CSS_GREY_BUTTON "QPushButton {" \
                            "color: #000000;" \
                            "background-color: #D3D3D3;" \
                            "padding: 12px;" \
                            "border: none;" \
                        "}" \
                        "QPushButton:hover {" \
                            "background-color: #DCDCDC;" \
                        "}" \
                        "QPushButton:pressed {" \
                            "color: #FFFFFF;" \
                            "background-color: #E57373;" \
                        "}"

#define CSS_BLUE_LINEEDIT   "QLineEdit {" \
                                "border: 2px solid #60B1C7;" \
                                "background: #cee8ef;" \
                                "selection-background-color: #3d96af;" \
                            "}" \
                            "QLineEdit:focus { border: 2px solid #3d96af; }"
#endif // COMMON_H

